### DISCUSSION TITLE
Tabs & tab management

### DISCUSSION THREAD

**Discussion by iamnbutler** [zed-industries/zed#6667] [Feature Requests]
I'd love to start a discussion on tab management. 
- Who does this well? 
- How should we deal with large amounts of tabs?
- Have you encountered any tab overflow experiences that are interesting or feel really good?
  - Example: Chrome's shrinking tabs, tab overflow dropdowns...
- How do you work with tabs? Do you try to close them as you go? Navigate with `cmd+p`?
- Are ephemeral tabs useful? (VSCode does this if you single click an item in the file tree for reference.)

Edit: Thanks for all the discussion y'all. I post these to let feedback accumulate so we can review them when we update these things. If they ever feel like nothing is happening based on all the feedback, it is likely we just haven't sequenced an update of that part of the editor. Reading through this one now 🙂

**Comment by JosephTLyons** [comment]
Tabs are a weird thing.  They are essential, but I've never found any one system that is great.  That being said, I have no idea what would make them any better.  

Some inspiration - Jetbrains' IDEs have a bunch of ways to tweak tab behavior

<img width="386" alt="image" src="https://user-images.githubusercontent.com/19867440/175788586-75c8c1e2-5cb9-422e-9c77-3580f6fb6c24.png">

<img width="320" alt="image" src="https://user-images.githubusercontent.com/19867440/175788604-f21fcb3d-4d5b-4983-94e3-bec176b7db6c.png">

The items under "Closing Policy" are basically a more advancec [`Zen Tabs`](https://github.com/zed-industries/zed/issues/1028) mode.  I'm not sure everyone likes that sort of thing; it doesn't really solve what to do if the setting is disabled.

Jetbrain's also allows of the tab bar to be placed vertically, which lets you view a lot more at once, since the amount that fits on the screen isn't determined by the amount of text, but the static height of the tab button itself.  Not everyone's cup of tea, just a different way to handle things.

**Reply by iamnbutler** [reply]
Thanks! Some of these settings are really interesting.

**Reply by elgerlambert** [reply]
> Jetbrain's also allows of the tab bar to be placed vertically, which lets you view a lot more at once

I know this thread and this comment is old, but this remark (still) hits the nail on the head for me! I really prefer the overview that a vertical tab bar provides!

I've seen other editors do the "active files above the file-tree" thing that was also suggested by @rkusa below, but I find this far to cluttered/distracting - I like to keep my file-tree panel closed most of the time.

**Comment by rkusa** [comment]
Propably controversial, but I personally like tabs to be reserved for projects (or repositories) and don't use them for files. I have this e.g. in VSCode and Sublime on macOS, where the native macOS tabs are used for projects/repos opened, and the VSCode/Sublime built-in tabs for files (doesn't work that way on Windows though).

If tabs are project/repository only, I never end up with too many open tabs in parallel. So tab management isn't an issue in this case.

For files, I consider the following good alternatives for tabs:
a) a vertical list of active files above the file-tree ([example](https://web.archive.org/web/20220416210929im_/https://chocolatapp.com/13/screenshots/splits.png))
b) CTRL+P and a command similar to CTRL+P that only lists/searches active files
c) two variants of CTRL+I / CTRL-O (navigate the history where the cursor has been): one to navigate the jump points in the current file; and another one to jump back and forth between the file history (and in each file to the last jump point in that file)

I still use tabs in Sublime and VSCode, mostly because (c) is lacking in both. I usually just close everything and start fresh when I start work on a new feature/bug - so not really a good workflow.

I've disabled ephemeral tabs as I never was able to get used to them and always found it confusing when the file was gone after navigation away.

**Reply by iamnbutler** [reply]
I think we will have some sort of project switcher, not sure if it will take the form of tabs or something else.

a) Would you always want this to be tied to the file tree (not for example, it's own panel)?
b) We have cmd + P, the second one is interesting. New keybinds are hard, but could be something like cmd+opt+p
c) We have history, but I think it combines these two things. Is having these two navigation types separate something you feel like you need?

**Reply by JosephTLyons** [reply]
> ... We have cmd + P, the second one is interesting. 

Both VS Code and Sublime Text show the open files directly in the normal file browser, sorted to the top of the list.

**Reply by diktomat** [reply]
> I think we will have some sort of project switcher, not sure if it will take the form of tabs or something else.

I'd like a project switcher to auto-discover projects by having it look for git repos inside a directory. Like, I have `~/dev` set as project containing root, and every subdirectory containing a git (or hg/bzr/...) repo automatically is picked up by the project switcher.

**Comment by JosephTLyons** [comment]
> a) a vertical list of active files above the file-tree ([example](https://web.archive.org/web/20220416210929im_/https://chocolatapp.com/13/screenshots/splits.png))

I do like the vertical list of active files too.  What editor are you showing here?

**Reply by jjlorenzo** [reply]
> I do like the vertical list of active files too. What editor are you showing here?

https://chocolatapp.com/ (I think is discontinued)

**Reply by TerminalFi** [reply]
Sublime also has this capability and I use it frequently

**Comment by exalted** [comment]
Man! Tabs, from a UX/UI standpoint, are tricky!

If you think about it, for dominant keyboard usage (see "code editors"), they don't add much value IMO, other than telling you "these are some of the files you've opened recently". At least in my personal usage, I'd navigate through code/files by other means anyway (e.e., "go to definition", `Ctrl-tab`'ing, etc.)

There is a case however that makes tabs somewhat useful IMO, however: Visual hints that show whether a file contains changes that need to be saved. Perhaps vertical space taken by tabs might be better used by integrating more such hints.

On the other hand, for dominantly mouse users, I guess tabs may sense as a means to faster access to… recently opened files. 😉 

Addressing some of @iamnbutler 's specific questions below:

> How do you work with tabs? Do you try to close them as you go? Navigate with `cmd+p`?

Often times I feel the urge to close all open tabs with: `Cmd-Alt-w` (will close all but the current tab) + `Cmd-w` (will close the current tab). (Why don't I have a shortcut to close'em all in one go? I don't have a good answer to that! 😅 )

> Are ephemeral tabs useful? (VSCode does this if you single-click an item in the file tree for reference.)

Absolutely! I mean in my ideal world that feature shouldn't even exist as it wouldn't be needed in the first place, but as I tried to hint at it above somewhat, I hardly ever reach out to (or interact with) my [open] tabs via a mouse, so they're 99% worthless to me (except for the visual hints they may provide). If any files that I "peek into" were to open in a new tab, I'd go crazy, I think.

So to recap, I guess my 2c about tabs is this: Does Zed want to provide a file-centric editor UX or is it more about the entire codebase regardless of how it's split across the filesystem. If you think about it, the editable multi-file buffer view which lets you see errors, search results, etc. does, in my mind, crosses a peculiar line where it isn't as important as where a given piece of code is saved. From that perspective, I think in the long term Zed could take some more bold decisions around what to do about tabs, more than "merely" showing recently opened files.

I saw an "HTML/CSS editor" ~~which I cannot recall its name on top of my head~~ (https://brackets.io) that showed relevant CSS class definitions _within_ the HTML. So, rather than opening the CSS file that contained, say a `.foo` class, on a separate tab, it would instead show you the relevant portion of the `*.css` file containing `.foo` right below the HTML code. I thought that was neat! When I am talking about "bold decisions", that's what I am thinking in terms of "re-thinking" tabs, et al. Imagine "peeking into" (and being able to edit) functions from the calling point, without ever leaving the "file" you're currently working in.

(I guess it would be only fair to make a similar discussion also for the tree viewer, but I will leave that for another day 😄 )

✌️

**Reply by iamnbutler** [reply]
Inline editors are very interesting. We have some ideas around ephemeral views into parts of files that we are kicking around right now... It would be interesting to look at all the different ways this has been done.

**Reply by iamnbutler** [reply]
Would you disable tabs if you had the option? The state we show in the tab today could easily move down to the breadcrumb.

> If any files that I "peek into" were to open in a new tab, I'd go crazy, I think.

I think Zed works this way today right?  I don't think that is desirable, but I think it does work this way right now.

**Reply by exalted** [reply]
> Would you disable tabs if you had the option? The state we show in the tab today could easily move down to the breadcrumb.

Yes, I think for the most part I'd disable tabs, but at the same time, I wouldn't like to lose some hints about recent files that I made some changes.

> > If any files that I "peek into" were to open in a new tab, I'd go crazy, I think.
> 
> I think Zed works this way today right? I don't think that is desirable, but I think it does work this way right now.

Any file you'll click on the tree view (left panel) will open that file in a new tab, regardless if it was an accidental click, or did you actually mean to open that file.

**Reply by lougreenwood** [reply]
> If you think about it, for dominant keyboard usage (see "code editors"), they don't add much value IMO, other than telling you "these are some of the files you've opened recently". At least in my personal usage, I'd navigate through code/files by other means anyway (e.e., "go to definition", Ctrl-tab'ing, etc.)

I disagree, I use tabs extensively in vim. Specifically I use tabs to group related files - I'm a frontend developer and the framework which I use means that there are multiple files for a given "entity", for example, given a web page URL (aka a `route`), I have the following files:

- route.ts
- controller.ts
- template.hbs
- route-acceptance-test.js
- route-page-object.js

When I'm working on a feature, I like to have a tab per "entity", I will use a tool like `telescope-alternate` (vim) or `related files hopper` (vscode) to allow me to open all of the sibling files when I'm viewing one of these files, I'll then open them in splits in that tab (and use a "split auto-collapse accordion thingy" addon, to allow the main file to have focus and the other splits take up minimum space, but maintain a consistent order on screen).

Effectively I use tabs as something similar to folder organisation for disparate files which are related, but scattered amongst the file system.

I also make extensive use of `tabdrop` in vim so that when I attempt to open a file (go to definition etc) when I'm not viewing the destiniation file (but it is already open in some other tab), then vim will jump to the target tab and buffer - avoiding having the same file open in many "buffers" across many tabs.

So for me, I want tabs to allow multiple buffers (the vim way), not tabs to be separate buffers (the vscode way).

**Reply by mbylstra** [reply]
> I disagree, I use tabs extensively in vim. Specifically I use tabs to group related files - I'm a frontend developer and the framework which I use means that there are multiple files for a given "entity", for example, given a web page URL (aka a route), I have the following files:

This is similar to how I work in tmux (tabs for different domains, splits for different things within the same domain), and while I've never done it like that vim for code (never quite figured out a good way to do it), I think you've described ideally how I'd be doing it!

**Comment by aexvir** [comment]
> How do you work with tabs? Navigate with cmd+p?

tabs for me are definitely a nuisance \
if I want a list of files I already have the file explorer, and that should already show it's state (new, updated, etc)
for switching between buffers I much prefer the command palette experience, (cmd+p), where I can search for exactly the file I need instead of glancing through the tab bar and clicking on a small box

I'd much rather have those few px of vertical space used for code than for a tab bar that I'll never use 😅 

> Do you try to close them as you go?

not really, I never worry about how many buffers I have opened \
as long as the "file switcher" or whatever is called the cmd+p palette is able to sort the files by how recently I opened them that's good for me

personally I love the cmd+p+p shortcut of vscode where I can jump back and forth between two files \
oh, and also please allow opening files on a split when inputting cmd+enter or shift+enter from the file switcher

> Are ephemeral tabs useful? (VSCode does this if you single click an item in the file tree for reference.)

nope, it's one of the most annoying features I've ever experienced 😅 \
luckily I can just turn it off

> Would you disable tabs if you had the option?

yes, [please](https://github.com/zed-industries/feedback/issues/183) 😅

**Reply by iamnbutler** [reply]
> oh, and also please allow opening files on a split when inputting cmd+enter or shift+enter from the file switcher

Open an issue for this if one doesn't exist already!

Thanks for your thoughts!

**Reply by JosephTLyons** [reply]
> oh, and also please allow opening files on a split when inputting cmd+enter or shift+enter from the file switcher

- https://github.com/zed-industries/feedback/issues/54

**Reply by aexvir** [reply]
oh, nice 🙂 thanks joseph!

**Reply by mateoroldos** [reply]
I came into this discussion and was going to comment exactly the same thing.
The best tabs management solution is no tabs 🧘‍♂️

But to work without tabs, sorting the file switcher by last opened is a must.

**Comment by MultisampledNight** [comment]
(moved from Discord) Haven't used Zed even once yet, but dropping my thoughts after reading through the stuff in there:
- I'm using NeoVim, and in general I try to avoid tabs as much as possible. If I already have a file opened, I want to see it in parallel to the current one, if not, I can usually access it through looking in the file hierarchy. And in that file hierarchy (which is nnn) tabs are very useful, but they're merely a replacement of "bookmarks" of important folders.
- Large amounts of tabs could be handled by making tabs more of a "recently-used" view instead, which could be searched for both filename and critical file structures/contents, and show where the cursor was before leaving the file. Then opening a "tab" would require at least 4 actions assuming a mouse, searching for the "tab" button, clicking it, searching for interesting content + scrolling, clicking again, but that way the editor wouldn't run out of horizontal space.
- The only real case I work with tabs is more in the browser domain, and there I primarily move with Alt + tab number, close with Ctrl + W and open new tabs with Ctrl + T, barely using the mouse there

**Comment by widersky** [comment]
> Who does this well?

It seems to me that VSCode does it quite well, but not perfectly.

> How should we deal with large amounts of tabs?

In my opinion, the cards should "sway" to the width of the file name by default. As a programmer, I want to see where I have an open file, the names can not be hidden from me. For the same reason, I think that when there is no place in the card bar, the cards should be displayed in several rows.

> Have you encountered any tab overflow experiences that are interesting or feel really good?
> Example: Chrome's shrinking tabs, tab overflow dropdowns...

Squeezing cards is ok, but as I wrote above - file names should not hide.

> How do you work with tabs? Do you try to close them as you go? Navigate with cmd+p?

I try to close cards that I haven't used in a long time on an ongoing basis.

> Are ephemeral tabs useful? (VSCode does this if you single click an item in the file tree for reference.)

Previewing the file after a single click is very helpful.

**Comment by exalted** [comment]
I don't necessarily need/want this; however, how about closing open tabs after a certain amount of time if they weren't modified/opened?

**Reply by widersky** [reply]
This could be distracting when the tabs close during our work.

**Reply by iamnbutler** [reply]
I'd love an option for stale tabs to dismiss assuming we could do this in a way that would ensure your current tab _never_ shifts.

**Comment by photex** [comment]
Tabs take up a lot of visual space and are distracting to me. I turn off tabs on all my editors. (exception being Xcode and Visual Studio since I don't think I *can* turn them off there).

I don't think tabs have ever helped me much. I realized this when I spent a lot of time with Emacs (with the projectile and helm plugins). Being able to bring up a list of "open buffers" just as easily as bringing up a list of all project files, is fundamentally the thing I need instead of tabs.

What's worse for me as well are how Xcode and Visual Studio have "preview tabs". If I'm debugging or trying to understand a very large project, I'll be hopping all over the place. But they'll happily just close files I've visited which completely interrupts the flow of what I'm doing.

At $dayjob I'm working in a 6M line C++ codebase. I need fluent movement through the code more than I need tabs. Bookmarks, open buffer list, etc.

**Comment by LeakyBucket** [comment]
I will say one thing that I've often found myself wanting to do is group a set of buffers/tabs together so I can easily switch focus between related sets in my visible panes. I will often find that I want to look at a template and the controller/handler code for it along with a model or similar as a set, and in some cases switch between two different representations/clusters of those things to compare or just look at histories as a set too.

**Comment by jamie** [comment]
The best UX I've ever had for dealing with open files came via a jEdit plugin, which would render a tree view filtered to just the open files (and then I'd hide the tab bar). Combined with dynamically collapsing the directory tree where possible (ie, if you've only got one file open foo/bar/baz.txt, there's a directory entry for `foo/bar`), this adds organization to my "working state" of relevant files for whatever feature I'm working at, without having to deal with horizontal scrolling 10 files when my editor window can only show 4 at a time. Given that, Cmd+P, search in project, and some basic jump-to-definition and jump-to-test-file functionality, I wouldn't actually need the file listing all too often (maybe browsing a list of models or something for a thing whose name I forget?).

I recently rediscovered this kicking the tires in RubyMine (just change the Project Files to Open Files in the dropdown top left and tweak a few settings to keep all folders expanded by default), it looks like this and is glorious:
![rubymine-open-files](https://user-images.githubusercontent.com/1093/235267129-b0cd46b3-e006-4f61-8a28-079da6999a07.png)

**Comment by 0x2CA** [comment]
I think vim's design is pretty good, tabs windows buffers

**Reply by 0x2CA** [reply]
We need https://github.com/akinsho/bufferline.nvim

**Comment by synecdokey** [comment]
The only option I truly desire with tabs is to be able to hide them. Vertical space is at a premium, I navigate with the fuzzy finder, and I gather most of my context via splits, so tabs just add noise. They should probably do more than being able to be hidden, but I would be ill-placed to talk about it. 😞

**Comment by Joxter** [comment]
Hi, I'm a experienced WebStorm user. I'm looking for an alternative as my main code editor and I'm trying Zed right now. 

My best idea last year for tabs was hiding tabs at all. Most of the time I need: search (shift-shift), "Go to defenition"/"Find usages", recent files/places. And I don't mess with tabs any more.

Even project files (list of all files) is hidden most of the time.

**Comment by CurbaiCode** [comment]
> * How should we deal with large amounts of tabs?
> * Have you encountered any tab overflow experiences that are interesting or feel really good?

Answering both, I know [Vivaldi](https://vivaldi.com/) has a lot tab customization options. Though I'm not sure how it could/should be used for file navigation, I like the idea of grouping tabs together for large amounts of tabs. Also, having them in a sidebar (maybe above the project panel like in VSCode #5240), and the option for using breadcrumbs instead (like [iamnbutler said](https://github.com/zed-industries/zed/discussions/6667#discussioncomment-3122211)).

> * How do you work with tabs? Do you try to close them as you go? Navigate with cmd+p?
> * Are ephemeral tabs useful? (VSCode does this if you single click an item in the file tree for reference.)

Usually, I only keep the ones I am currently working in or hopping back-and-forth from open, closing them with the mouse (the close button could be a little bigger though). Sometimes I might leave one or two open for reference though, so I don't have to use the project panel and make ephemeral tabs.

Lastly, ~I don't remember who, but someone~ [jvanmourik](https://github.com/jvanmourik) proposed the idea of combining the tab bar and the title bar #5066, which is kind of like Safari's compact tab layout which I am a fan of.

**Comment by e1turin** [comment]
I like vim way with list of active buffers. And I like more the evolution of this idea in Neovim's plugin [Telescope](https://github.com/nvim-telescope/telescope.nvim) with its ability to show search through a list of buffers (built-in vim picker for buffer list) and preview their contents. So I do not need any tabs and just work with opened files. This way gives the ability to show additional information about a buffer, such as a git status and previous buffer, and I've created a keyboard shortcut for it ― hit space twice ― really convenient.

It's matter to say that Telescope gives huge ability for customization and integration with different subsystems, e.g. file system manipulations, git, grep, a search for editor settings, linter issues, file-bookmarks and so on. 

But I see that such behavior is not intuitive for beginners, and maybe it's good idea to place such buffer list not in popup but sidebar so it can be used without keyboard shortcut.

Moreover, I use my web-browser sidebar for tab tree (Sidebery extension) as it is more convenient to see a hierarchy and increase amount of displayed tabs. 

I hope these ideas will lead you to the most orthogonal solution to the tab navigation problem. ☕

**Reply by npckenjamin** [reply]
Agree with this! 
The main reason I still use neovim over zed right now, is the fact that panes get their own tabs/buffers.

A [bufferline-esque](https://github.com/akinsho/bufferline.nvim) feature could be good for beginners where the list of active buffers is shown as tabs and clicking on one only updates the active pane

**Comment by olevekavitocz** [comment]
Please make it possible to disable tabs completely. I use inteliJs products - phpstrom, and I completely get by with search, split view and last opened files. The whole use of tabs makes me want to work with the mouse, which I try to avoid :-)

**Reply by mbylstra** [reply]
I totally agree! Before this can be done though, it needs to be possible to open a file in a new split (horizontally or vertically), ideally with whatever key bindings you desire, from any of the file search or selection windows/panes. Seems there is currently only support for splitting vertically (with cmd-enter) and it's not possible to assign a different key.

**Comment by illustrofia** [comment]
I sometimes work with multiple tabs open and I love the ability to switch to the last used tab (how VSCode does it).

**Comment by liam-macann** [comment]
Coming from Sublime, I love being able to trigger split view with modifier keys. Just like in a list of files, you can open consecutive tabs with shift-click and select specific tabs with cmd-click. Works super well with the list of active files in the side bar, as well.

**Comment by piotrkowalczuk** [comment]
I would love to see the ability to turn off tabs completely.

**Reply by abejfehr** [reply]
Just curious: what would happen in that case if you selected a different file than the one you were on?

What if you had unsaved changes?

**Reply by piotrkowalczuk** [reply]
Currently, I use this setup in Goland (IntelliJ). Goland autosaves frequently and keeps local history. So, moving between files is just replacing what's being shown. Navigating using `cmd+[` or `cmd+]` naturally replaces what's rendered the same way as choosing a file from the project tree. 

Tabs settings:
<img width="656" alt="image" src="https://github.com/zed-industries/zed/assets/2438924/193cde33-15a9-43ec-9700-afbfd006fb11">

Autosave settings:
<img width="679" alt="image" src="https://github.com/zed-industries/zed/assets/2438924/ead0d69b-7ed5-4756-8360-fff5cae49084">

**Comment by jerrygreen** [comment]
Cycle though last used tabs, instead of just visual order, - is really important to me.
Isn't it possible yet? I couldn't find such a keymap setting.

**Reply by alygin** [reply]
It's implemented as the [Tab Switcher feature](https://github.com/zed-industries/zed/pull/7987). Will be released in a couple of weeks.

**Comment by sommerper** [comment]
I would like to see an option to move tabs.
Something like: 
```
"alt-i": "pane::GoBack",
"alt-n": "pane::GoForward",
"alt-shift-n": "pane::MoveBack",
"alt-shift-i": "pane::MoveForward"
```

**Reply by jerrygreen** [reply]
Moving tabs with a hotkey is one of the first things I found aren't accessible in Zed.

I'm so used to Ctrl-Shift-PgUp and Crtl-Shift-PgDown in browsers when using tabs.

**Comment by omid** [comment]
Regarding the vertical tabs, I'm just playing around with the source code to get more familiar with it.
It's what I have so far, but I'm not sure if the zed devs will accept the changes or not, or how generally the process (I've read the CONTRIBUTING file) works? It's not a tiny change, it's also not a big one!

So far, it's not resizable, and it's just _top_ or _right_ (and _none_, to remove one config?).
It's not a `pane`, it's exactly the same `tab_bar` we have on top.

Ref: https://github.com/zed-industries/zed/issues/4331 (and a bit https://github.com/zed-industries/zed/issues/5240)

![image](https://github.com/user-attachments/assets/e9ee2082-a85b-447e-951f-be63a710bfbc)

PS: the code is not even close to ready to be a PR.

**Reply by exalted** [reply]
@omid Hey, great stuff! Would you mind creating a separate issue (or PR) to discuss specifics of the changes in there please? You can mention this discussione from there, to "tie" the contexts together. Thanks!

**Reply by omid** [reply]
@exalted thanks, to avoid duplications, can i go with #4331?

**Reply by exalted** [reply]
No strong opinion on my end. I trust your decision! :high-five:

**Comment by psiho** [comment]
My 2c, and how I do it in Nvim:
- tabs are minimal (visually), and all splits share the same tab bar. Mostly I use them to arrange files in vertical splits
- I use Telescope to quickly search/preview/switch-to open tabs (buffers), with shortcut <leader>bb
- bp - go back to last active tab in tab history (bn - go next)
- <leader>bc - close tab
- <leader>bo - close al tabs and splits except active one

**Comment by gldtn** [comment]
Just to add to this, we could have an option to hide tabs for those that don't use it, and add a picker(fzf/telescope style) similar to cmd+shift+p(command palette) where we can see all the buffers/tabs open. This would mimic nicely for those who are used to nvim ways and don't use bufferline.

**Reply by Gambloide** [reply]
I use the setting 
```
"tab_bar": {
     "show": false
}
````
and use the keybinds for `tab_switcher::Toggle` and `file_finder::Toggle` to almost get what you ask for.

**However**, this comes with the major issue that https://github.com/zed-industries/zed/pull/54 and https://github.com/zed-industries/zed/pull/52 do not consider people having hidden their tab bar. So you could be editing a file for a good while after it was changed outside of zed and be oblivious to the fact, because the indicator for this is only in the tab an nowhere else.

IMO tab-bars in a text-editor are an accessibility feature and/or training wheels, and hence should not be a dependency for other features.

**Reply by jon-walstedt** [reply]
For me having the `tab_bar.show` setting to false and have a fzf like way of searching open buffers would be perfect. I much like the way vim handles this where you have windows that contains buffers, you can have any number of windows (splits) opened and have different buffers (or the same) open within each window.

**Comment by lknite** [comment]
Why not implement the tab group style of chrome/edge as the default, what everyone has come to know, and then add enhancements after that for all these unique use cases.  Chrome/edge style tabs would switch me from code to zed.  Its been a requested feature over there for 4 years: https://github.com/microsoft/vscode/issues/100335

**Comment by grrowl** [comment]
Would love at least a customisable maxWidth option for tabs, or specifically pinned tabs — this is especially useful for Center Terminal tabs where they can take up half the tab bar on a horizontally split window

**Comment by mirql** [comment]
maybe we can make tab works like in neovim? :3
buffers can be in any tabs/windows, so tabs is just a "schema for pane/window-splits config"

**Comment by lionel-** [comment]
Another source of inspiration is the Firefox extension Sidebery: https://github.com/mbnuqw/sidebery

It implements tree-like vertical tabs where you can group tabs together by nesting them. You can fold or close entire nodes, it's very convenient.

**Reply by lionel-** [reply]
Just want to add that this Firefox extension is worth looking at for all its little usability gems. For instance clicking on a tab switches you to the last selected tab, which is very handy for back and forth toggling (click on a tab to switch to it, click again to switch back to where you started). Or you can create a new tab group by dragging a bunch of selected tabs to a `+` icon.

**Comment by imsaravana369** [comment]
Any plans to add support for [native tabs](https://lucasprag.com/posts/underrated-vscode-feature-native-tabs/#:~:text=How%20to%20setup%20Native%20Tabs,json%20.) like in VS Code? I find it super helpful when working on multiple projects at the same time

**Reply by testpassword** [reply]
+++. For me, this is the last thing that is missing to fully switch from JetBrains products

**Reply by elalecs** [reply]
You can set a workspace with arbitrary folders, aka vertical native tabs. Doesn't that meet your need?

<img width="1039" height="587" alt="image" src="https://github.com/user-attachments/assets/7dca4d43-8339-4039-b092-504aedecde39" />

<img width="1039" height="587" alt="image" src="https://github.com/user-attachments/assets/25bcd613-8ffe-454f-b9a7-4bd6c1b94992" />

<img width="1039" height="587" alt="image" src="https://github.com/user-attachments/assets/7e9dadf2-588d-4ea1-884a-a67820e3d55d" />

**Comment by lionel-** [comment]
A huge UX leap for Zed would be "tabs of tabs", or "grouped tabs".

Within a single project I often work with totally different sets of tabs/files depending on the branch I am on or the task I'm doing. I would love to be able to organise different sets of tabs for each context.

Example from Sidebery (each icon in the top row is a group of tabs, the second row shows pinned tabs in that group):


https://github.com/user-attachments/assets/d92457e9-d601-4fe9-9791-047256568c2d

Note most of these tabs are "unloaded" so they don't take up memory/cycles when I'm not using them (that's why they are grayed out). The same concept _might_ be relevant in an IDE for language servers, to avoid adding unused files opened in tabs to the set of opened files managed by the LSP if there is no need. You can unload tabs individually or by group via the context menus.

**Reply by masrurimz** [reply]
This really good, also, i would think the UX will be more like Arc browser.
instead of navigation primary is file explorer, the primary is vertical grouped tab (also included panel), we can treat it just like a panel, or an essentials tab just like arc
so we can have multiple grouped set-of tab+panel, which weill really good to be able switch into diffrenet workflow and stash/store panel configuration if another task coming

**Comment by elalecs** [comment]
When I'm working with multiple files at once, my workflow involves using panels to organize files and terminals. There are files that I need to keep pinned so I don't lose them, which causes the tab bar to limit the available space for displaying non-pinned tabs.

<img width="1213" height="843" alt="image" src="https://github.com/user-attachments/assets/8bcd1e97-0970-421b-ac73-8b2ddc79142a" />

What I implemented is bringing the solution from VS Code: being able to show pinned tabs in another row and/or change the size of the fixed tabs.

<img width="1213" height="844" alt="image" src="https://github.com/user-attachments/assets/06808bad-035b-4e72-9580-103c5c4572ef" />

<img width="1213" height="844" alt="image" src="https://github.com/user-attachments/assets/48cf66d6-98a8-4d87-a870-c4e1a5de2637" />

<img width="1213" height="844" alt="image" src="https://github.com/user-attachments/assets/b4ef84f6-6a6e-458d-be86-446cdf3f8697" />

<img width="1213" height="844" alt="image" src="https://github.com/user-attachments/assets/b04cecf1-82e3-4a50-a6ae-1c9bb8f17491" />

Not having these mechanisms makes working with multiple panels and multiple files at once tedious.

**Reply by Red-Devil-SDO** [reply]
Also voting for this. Because for me, I very sorrow, that Zed miss this functionality. I often open a lot of tabs, and switch between them, just have one row - it's very uncomfortable and hard. It would be great, if in Zed implements multi-row functionality for opened tabs with files. Or if it very complicated, just add some list with opened files. It would be great if you add this functionality. Thank you.

**Reply by elalecs** [reply]
This is already implemented since July but team doesn't allow to merge it, hope they change their mind

**Reply by yaroslavrick** [reply]
Hello @elalecs I've seen your [PR](https://github.com/zed-industries/zed/pull/34469) - great job made and many interesting features implemented there! It is sad that your PR was closed..
It seems that it was closed becase to many features in one PR.

I've implemented in my [PR](https://github.com/zed-industries/zed/pull/46573) feature 'show pinned tabs in a separate row '. This PR is merged to master branch and is released in 'Zed Preview v0.220.0-pre' - you can use it already:

<img width="3024" height="318" alt="image" src="https://github.com/user-attachments/assets/bc79c9da-dd6b-452e-95da-30cff8935511" />

But in your PR there are many other interesting features were implemented, it would be great if you implement them separately in separate PRs - there is a chance that Zed team will accept all of these changes:

- Implements three pinned tab sizing modes: normal, shrink, and compact
- Introduces pane locking to prevent accidental modifications
- Adds maximum pinned tabs limit with user feedback
- Implements smart item routing when panes are locked
- etc

**Comment by antonr-p2p** [comment]
I wanted to switch from VSC to Zed because of painful VSC tabs managing, but Zed unfortunately lacks this as well. Why this is such a problem in 2025 to work with multiple projects in just one IDE window ??? All modern browser manage to group tabs, what is wrong with IDEs?

**Comment by jakobolason** [comment]
What about making a timeout on window lifetimes, closing them down after a selected amount of time, like 30mins? I like tab+alt to switch between recently-used buffers, and then ctrl+p to find other files, but I'm horrible at closing tabs down, so they just pile up. A bit inspired by the Arc browser when that was good..

**Comment by YaLTeR** [comment]
I don't mind tabs, but there's one annoying behavior in vim mode. I frequently use the "buffer picker" via Zed's closest approximation of `tab_switcher::ToggleAll`, but it always opens the file in the pane where it was already open, instead of the current pane.

For example, if I have a two-pane split like this:

```
┌─────────────┬─────────────┐
│ [A] [B]*    │ [C]         │
├─────────────┼─────────────┤
│             │             │
│      B      │      C      │
│             │             │
└─────────────┴─────────────┘
       ^ focus
```

After I open the "buffer picker" and select C:

```
┌─────────────┬─────────────┐
│ [A] [B]     │ [C]*        │
├─────────────┼─────────────┤
│             │             │
│      B      │      C      │
│             │             │
└─────────────┴─────────────┘
                     ^ focus jumps here
```

What I want instead:

```
┌─────────────┬─────────────┐
│ [A] [B] [C]*│ [C]         │
├─────────────┼─────────────┤
│             │             │
│      C      │      C      │
│             │             │
└─────────────┴─────────────┘
       ^ focus stays here
```

This happens even if the tab bar is hidden in the settings.

**Reply by baldwindavid** [reply]
#47079 adds a mode to `tab_switcher::ToggleAll` that opens files in whichever pane is active.

**Comment by webmasterMeyers** [comment]
all I want is new tabs/editors to open on the far right, the END of the tab row.

**Comment by timkindberg** [comment]
WRAP TABS option PLS, you should have to scroll horizontally to see all your tabs, they should just wrap to a new line (as an option)

**Comment by ProCode-Software** [comment]
I want collapsible tab groups, similar to the ones in Chrome, so I can categorize the files I'm working on and switch between them
